File: src/modules/auth/services/auth-register.service.ts
import { ConflictException, Injectable } from '@nestjs/common';
import { $Enums } from '@prisma/client';
import { hash } from 'bcryptjs';
import { PrismaService } from 'src/infra/database/prisma.service';
import { UserEntity } from 'src/modules/user/domain/entities/user.entity';
import { AuthRegisterDto } from '../domain/dto/auth-register.dto';
import { EMailTemplate } from 'src/modules/mail/domain/enums/mail-template.enum';
import { IMailOption } from 'src/modules/mail/domain/interfaces/mail-option.interface';
import { MailService } from 'src/modules/mail/services/mail.service';
import { IAuthOrganizationActivation } from '../domain/interfaces/auth-organization-activation.interface';

@Injectable()
export class AuthRegisterService {
  constructor(
    private prisma: PrismaService,
    private mailService: MailService,
  ) {}
  async execute(dto: AuthRegisterDto): Promise<UserEntity> {
    const isAdminOrganization: boolean =
      dto.reference === $Enums.RoleReferences.ADMIN_ORGANIZATION;

    const user = await this.prisma.user.findFirst({
      where: {
        OR: [{ name: dto.name }, { email: dto.email }],
        deletedAt: null,
      },
    });

    if (user)
      throw new ConflictException('Nome ou e-mail de Usuário já cadastrado!');

    const organization = await this.prisma.organization.findFirst({
      where: {
        OR: [{ name: dto.organizationName }, { email: dto.organizationEmail }],
        deletedAt: null,
      },
    });
    if (organization)
      throw new ConflictException(
        'Nome ou e-mail da organização já cadastrado!',
      );

    const createdUser = await this.prisma.user.create({
      data: {
        name: dto.name,
        email: dto.email,
        gender: dto.gender,
        phone: dto.phone,
        status: isAdminOrganization
          ? $Enums.Status.WAITING
          : $Enums.Status.ACTIVE,
        password: await hash(dto.password, 10),
        role: {
          connect: {
            id: isAdminOrganization ? 2 : 4,
          },
        },
        organization: isAdminOrganization
          ? {
              create: {
                name: dto.organizationName,
                tradeName: dto.organizationTradeName,
                email: dto.organizationEmail,
                status: $Enums.Status.WAITING,
                address: {
                  create: dto.address,
                },
              },
            }
          : undefined,
      },
      include: {
        role: true,
        organization: true,
        company: true,
        avatar: true,
      },
    });

    if (isAdminOrganization) {
      const options: IMailOption<IAuthOrganizationActivation> = {
        to: process.env.ADMIN_EMAIL,
        subject: 'Cadastro de organização:',
        template: EMailTemplate.ORGANIZATION_ACTIVATION,
        context: {
          name: createdUser?.organization?.name,
          url: `${process.env.FRONT_END_URL}`,
        },
      };

      await this.mailService.send(options);
    }

    return new UserEntity(createdUser);
  }
}

File: src/modules/authentication/services/auth_register_service.py
import os
from django.db.models import Q
from src.modules.authentication.domain.dto.auth_register_dto import AuthRegisterDTO
from src.modules.user.models import User
from src.modules.role.models import Role
from src.modules.organization.models import Organization
from src.modules.address.models import Address
from src.shared.exceptions.conflict_exception import ConflictException
from src.shared.utils.hash import hash
from src.modules.role.domain.enums.role_references_enum import RoleReferences
from src.shared.domain.enums.status_enum import Status
from src.modules.mail.services.mail_service import MailService
from src.modules.mail.domain.enums.mail_template_enum import EmailTemplate
from src.modules.mail.domain.interfaces.mail_option_interface import IMailOption
from src.modules.authentication.domain.interfaces.auth_recover_password_interface import (
    IAuthRecoverPassword,
)

from src.modules.user.validators.user_password_validator import user_password_validator


class AuthRegisterService:
    def execute(dto: AuthRegisterDTO):
        is_admin_organization = dto.reference == RoleReferences.ADMIN_ORGANIZATION

        user = User.objects.filter(
            (Q(name=dto.name) | Q(email=dto.email)) & Q(deleted_at=None)
        ).first()

        if user:
            raise ConflictException("Nome ou e-mail de usuário já cadastrado!")

        if is_admin_organization:

            organization = Organization.objects.filter(
                (Q(name=dto.organization_name) | Q(email=dto.organization_email))
                & Q(deleted_at=None)
            ).first()

            if organization:
                raise ConflictException("Nome ou e-mail da organização já cadastrado!")

        user_password_validator(dto.password)

        createdUser = User.objects.create(
            name=dto.name,
            email=dto.email,
            gender=dto.gender,
            phone=dto.phone,
            status=Status.WAITING if is_admin_organization else Status.ACTIVE,
            password=hash(dto.password),
            role=Role.objects.get(id=2 if is_admin_organization else 4),
            organization=(
                Organization.objects.create(
                    name=dto.organization_name,
                    trade_name=dto.organization_trade_name,
                    email=dto.organization_email,
                    status=Status.WAITING,
                    address=Address.objects.create(
                        state=dto.address.state,
                        city=dto.address.city,
                        street=dto.address.street,
                        neighborhood=dto.address.neighborhood,
                        complement=dto.address.complement,
                    ),
                )
                if is_admin_organization
                else None
            ),
        )

        if is_admin_organization:
            context: IAuthRecoverPassword = {
                "name": createdUser.organization.name,
                "url": os.getenv("FRONT_END_URL", ""),
            }
            options: IMailOption = {
                "to": os.getenv("ADMIN_EMAIL", ""),
                "subject": "Cadastro de organização:",
                "template": EmailTemplate.ORGANIZATION_ACTIVATION,
                "context": context,
            }
            MailService.send(options=options)

        return createdUser
